server/linkpreview.go

```diff
@@ -0,0 +1,148 @@
+package main
+
+import (
+	"encoding/json"
+	"golang.org/x/net/html"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type linkPreview struct {
+	Title       string `json:"title,omitempty"`
+	Description string `json:"description,omitempty"`
+	ImageURL    string `json:"image_url,omitempty"`
+}
+
+var client = &http.Client{
+	Transport: &http.Transport{},
+	Timeout:   time.Second * 2,
+}
+
+// previewLink handles the HTTP request, fetches the URL, and returns the link preview.
+func previewLink(w http.ResponseWriter, r *http.Request) {
+	u := r.URL.Query().Get("url")
+	if u == "" {
+		http.Error(w, "Missing 'url' query parameter", http.StatusBadRequest)
+		return
+	}
+
+	parsedURL, err := url.Parse(u)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		http.Error(w, "invalid schema", http.StatusBadRequest)
+		return
+	}
+
```

Comment:
`http.Client` follows redirects. It means the first URL can look legit but then it would redirect to `http://localhost/`. Take a look at `Client.CheckRedirect`. I think URL validation should be moved to a separate function and called from  `CheckRedirect`.

Reply:
Or disable redirects altogether.